iroh-net/src/magic_endpoint.rs

```diff
@@ -0,0 +1,561 @@
+    callbacks: Callbacks,
+}
+
+impl MagicEndpointBuilder {
+    /// Set a keypair to authenticate with other peers.
+    ///
+    /// This keypair's public key will be the [PeerId] of this endpoint.
+    ///
+    /// If not set, a new keypair will be generated.
+    pub fn keypair(mut self, keypair: Keypair) -> Self {
+        self.keypair = Some(keypair);
+        self
+    }
+
+    /// Set the ALPN protocols that this endpoint will accept on incoming connections.
+    pub fn alpns(mut self, alpn_protocols: Vec<Vec<u8>>) -> Self {
+        self.alpn_protocols = alpn_protocols;
+        self
+    }
+
+    /// If *keylog* is `true` and the KEYLOGFILE environment variable is present it will be
+    /// considered a filename to which the TLS pre-master keys are logged.  This can be useful
+    /// to be able to decrypt captured traffic for debugging purposes.
+    pub fn keylog(mut self, keylog: bool) -> Self {
+        self.keylog = keylog;
+        self
+    }
+
+    /// Specify the DERP servers that are used by this endpoint.
+    pub fn derp_map(mut self, derp_map: Option<DerpMap>) -> Self {
+        self.derp_map = derp_map;
+        self
+    }
+
+    /// Set a custom [quinn::TransportConfig] for this endpoint.
+    pub fn transport_config(mut self, transport_config: quinn::TransportConfig) -> Self {
+        self.transport_config = Some(transport_config);
+        self
+    }
+
+    pub fn concurrent_connections(mut self, concurrent_connections: u32) -> Self {
+        self.concurrent_connections = Some(concurrent_connections);
+        self
+    }
+
+    /// Optionally set a callback function to be called when endpoints change.
+    #[allow(clippy::type_complexity)]
+    pub fn on_endpoints(
+        mut self,
+        on_endpoints: Box<dyn Fn(&[cfg::Endpoint]) + Send + Sync + 'static>,
+    ) -> Self {
+        self.callbacks.on_endpoints = Some(on_endpoints);
+        self
+    }
+
+    /// Optionally set a callback funcion to be called when a connection is made to a DERP server.
+    pub fn on_derp_active(mut self, on_derp_active: Box<dyn Fn() + Send + Sync + 'static>) -> Self {
+        self.callbacks.on_derp_active = Some(on_derp_active);
+        self
+    }
+
+    /// Optionally set a callback function that provides a [cfg::NetInfo] when discovered network conditions change.
+    pub fn on_net_info(
+        mut self,
+        on_net_info: Box<dyn Fn(cfg::NetInfo) + Send + Sync + 'static>,
+    ) -> Self {
+        self.callbacks.on_net_info = Some(on_net_info);
+        self
+    }
+
+    /// Bind the magic endpoint on the specified socket address.
+    ///
+    /// The *bind_addr* is the address that should be bound locally.  Even though this is an
+    /// outgoing connection a socket must be bound and this is explicit.  The main choice to
+    /// make here is the address family: IPv4 or IPv6.  Otherwise you normally bind to the
+    /// `UNSPECIFIED` address on port `0` thus allowing the kernel to do the right thing.
```

Comment:
Good question, this is taken over from the old `dial_peer` functions and the code in `iroh/src/node.rs`
Now that I re-read the code, the `bind_addr` IP address & family  is not even used, only its port is passed to the `magicsock::Conn`. So - it would actually be correct to remove the `bind_addr: SocketAddr` argument and instead add `bind_port: u16`?
@dignifiedquire can you confirm this? Then I'd go ahead and replace `bind_addr` with `bind_port`.

Reply:
My answer is two fold, the current state and the ideal state.

- Current State: The address is effectively ignored and only the port is used, as we bind to IPv4 and IPv6 seperately.
- Should State: Allow to provide sth like this 
```rust
enum BindAddress {
  /// Bind on this port, both ipv4 and ipv6 addr, using `0.0.0.0` and `[::]`
  Port(u16),
  Ips(Ips),
}
enum Ips {
  // Only Ipv4
  Ipv4(SocketV4Addr),
  // Only Ipv6
  Ipv6(SocketV6Addr),
  Both {
    ipv4: SocketV4Addr,
    ipv6: SocketV4Addr,
}

// The `BindAddress::Port` options is equivalent to 
BindAddress::Ips(Ips::Both {
  ipv4: SocketV4Addr::new(Ipv4Addr::UNSPECIFIED, port),
  ipv6: SocketV6Addr::new(Ipv6Addr::UNSPECIFIED, port),
});

// so could probably change this to just have `Ips` and this
impl Ips {
  pub fn from_port(port: u16) -> Self {
    Ips::Both {
      ipv4: SocketV4Addr::new(Ipv4Addr::UNSPECIFIED, port),
      ipv6: SocketV6Addr::new(Ipv6Addr::UNSPECIFIED, port),
    }
  }
}
```

Reply:
Yes, that makes more sense.  Maybe `Ips` should even be a vec, I think there's nothing wrong with binding to a bunch of specific addresses, e.g. you're on say 20 subnets but only want to bind to the private ones.

Reply:
So I'd be mildly in favour of for now changing this to `u16` and describe that this will change in the future to allow binding to specific socket addresses.

Reply:
> e.g. you're on say 20 subnets but only want to bind to the private ones

That would be a lot of internal changes, as we only do at most two sockets internally. I would argue, this should be managed by some network configuration/proxy infront if you have multiple private binds.